lib/fennec/evaluator/allocate/request.ex

```diff
@@ -0,0 +1,112 @@
+defmodule Fennec.Evaluator.Allocate.Request do
+  @moduledoc false
+
+  alias Jerboa.Format.Body.Attribute
+  alias Jerboa.Params
+  alias Fennec.TURN
+  @lifetime 10 * 60
+
+  @spec service(Params.t, map, TURN.t) :: {Params.t, TURN.t}
+  def service(x, changes, turn_state) do
+    request_status =
+      {:valid, x, %{}}
+      |> maybe(&verify_existing_allocation/4, [changes, turn_state])
+      |> maybe(&verify_requested_transport/2)
+      |> maybe(&verify_unknown_attributes/2)
+
+    case request_status do
+      {:error, error_code} ->
+        {%{x | attributes: [error_code]}, turn_state}
+      {:ok, {new_params, new_turn_state}} ->
+        {new_params, new_turn_state}
+      {:valid, _, state} ->
+        allocate(x, state, changes, turn_state)
+    end
+  end
+
+  defp allocate(x, _state, changes, turn_state) do
+    addr = Application.get_env(:fennec, :relay_addr, {127, 0, 0, 1})
+    {:ok, socket} = :gen_udp.open(0, [:binary, active: true, ip: addr])
+    allocation = %Fennec.TURN.Allocation{
+      socket: socket,
+      expire_at: System.system_time(:second) + @lifetime,
+      owner: Params.get_id(x)
+    }
+
+    new_turn_state = %{turn_state | allocation: allocation}
+    allocation_params(x, changes, new_turn_state)
+  end
+
+  defp allocation_params(x, %{address: a, port: p},
+                         turn_state = %TURN{allocation: allocation}) do
+    addr = Application.get_env(:fennec, :relay_addr, {127, 0, 0, 1})
+    %TURN.Allocation{socket: socket, expire_at: expire_at} = allocation
+    {:ok, port} = :inet.port(socket)
+    lifetime = max(0, expire_at - System.system_time(:second))
```

Comment:
Please not that this function is used both while allocating new allocation and for generating response for existing one.